src/passes/StripEH.cpp

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Remove catch blocks and convert 'throw's into 'unreachable's. Any exception
+// thrown will crash the program as they are now traps.
+//
+
+#include <ir/drop.h>
+#include <pass.h>
+#include <wasm-builder.h>
+#include <wasm.h>
+
+namespace wasm {
+
+struct StripEH : public WalkerPass<ExpressionStackWalker<StripEH>> {
+  bool isFunctionParallel() override { return true; }
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<StripEH>();
+  }
+
+  void visitThrow(Throw* curr) {
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+    replaceCurrent(getDroppedChildrenAndAppend(curr,
+                                               wasm,
+                                               getPassOptions(),
+                                               builder.makeUnreachable(),
+                                               DropMode::IgnoreParentEffects));
+  }
+
+  void visitTry(Try* curr) { replaceCurrent(curr->body); }
```

Comment:
Does the body always have the same type as the Try? If the Try has type`i32` and the body has type `unreachable` (possible with an If, not sure offhand about Try) then this will error as a refinalize is needed.

Reply:
Can you give an example of this that fails validation after this without refinalization? I can't really think of one. Now that the `try` is gone, we have its parent operation that expects an `i32` but rather gets `unreachable`, but that's fine, no? For example, 
```
(i32.add
  (try (result i32)
    (throw ...)
    ...
  )
  (i32.const 0)
)
```
Will become
```wast
(i32.add
  (unreachable)
  (i32.const 0)
)
```
And 
```wast
(block (result i32)
  (try (result i32)
    (throw ...)
    ...
  )
)
```
will become
```wast
(block (result i32)
  (unreachable)
)
```
and these seem to be OK.

In which case we need refinalization in general?

Reply:
In your first piece of code the type of the `i32.add` itself is `i32`, while in the second it should be `unreachable`. That is, after one of the add's children becomes unreachable, we need to update the parent.

With `BINARYEN_PRINT_FULL=1` those types are printed, and with `BINARYEN_PASS_DEBUG=1` the types are checked, so I think in that mode you'll get an error on that testcase.

In general, replacing something with an identical type is always ok with no additional work. Replacing something with another type - either unreachable, or a subtype - requires type updates in the parent (and perhaps further), which we need refinalize for.

Reply:
I see, fixed it and added a test for it.